openmmtools/alchemy.py

```diff
@@ -12,413 +12,905 @@
+
+    @classmethod
+    def _get_supported_parameters(cls):
+        """Return a set of the supported alchemical parameters.
+
+        This is based on the exposed properties. This ways we keep autocompletion
+        working and avoid silent bugs due to possible monkey patching caused by
+        a typo in the name of the variable.
+
+        """
+        # TODO just use inspect.getmembers when dropping Python 2
+        supported_parameters = {name for name, value in cls.__dict__.items()
+                                if isinstance(value, cls._LambdaProperty)}
+        return supported_parameters
+
+    @classmethod
+    def _get_system_lambda_parameters(cls, system):
+        """Yields the supported lambda parameters in the system.
+
+        Yields
+        ------
+        A tuple force, parameter_name, parameter_index for each supported
+        lambda parameter.
+
+        """
+        supported_parameters = cls._get_supported_parameters()
+
+        # Retrieve all the forces with global supported parameters.
+        for force_index in range(system.getNumForces()):
+            force = system.getForce(force_index)
+            try:
+                n_global_parameters = force.getNumGlobalParameters()
+            except AttributeError:
+                continue
+            for parameter_id in range(n_global_parameters):
+                parameter_name = force.getGlobalParameterName(parameter_id)
+                if parameter_name in supported_parameters:
+                    yield force, parameter_name, parameter_id
+
+
+# =============================================================================
+# ALCHEMICAL REGION
+# =============================================================================
+
+_ALCHEMICAL_REGION_ARGS = collections.OrderedDict(
+    [('alchemical_atoms', None), ('alchemical_bonds', None),
+     ('alchemical_angles', None), ('alchemical_torsions', None),
+     ('annihilate_electrostatics', True), ('annihilate_sterics', False),
+     ('softcore_alpha', 0.5), ('softcore_beta', 0.0), ('softcore_a', 1), ('softcore_b', 1),
+     ('softcore_c', 6), ('softcore_d', 1), ('softcore_e', 1), ('softcore_f', 2)]
+)
+
+
+# The class is just a way to document the namedtuple.
+class AlchemicalRegion(collections.namedtuple('AlchemicalRegion', _ALCHEMICAL_REGION_ARGS.keys())):
+    """Alchemical region.
+
+    This is a namedtuple used to tell the AlchemicalFactory which region
+    of the system to alchemically modify and how.
+
+    Parameters
+    ----------
+    alchemical_atoms : list of int, optional
+        List of atoms to be designated for which the nonbonded forces (both
+        sterics and electrostatics components) have to be alchemically
+        modified (default is None).
+    alchemical_bonds : bool or list of int, optional
+        If a list of bond indices are specified, these HarmonicBondForce
+        entries are softened with 'lambda_bonds'. If set to True, this list
```

Comment:
The argument handles both `None` and `False` (see line 915) since it can take both a list and a bool. I think setting this to `True` will be the most common way to use this, so I wouldn't raise warning. I can change the default to `False` if you think it's better.